python/ray/tune/automl/genetic_searcher.py

```diff
@@ -0,0 +1,206 @@
+# date: 2018/06/27
+
+import logging
+import numpy as np
+
+from .search_policy import AutoMLSearcher
+
+
+class GeneticSearch(AutoMLSearcher):
+    """Implement the genetic search.
+
+    Keep a collection of top-K parameter permutations as base genes,
+    then apply selection, crossover, and mutation to them to generate
+    new genes (a.k.a new generation). Hopefully, the performance of
+    the top population would increase generation by generation.
+    """
+
+    def __init__(self, search_space, reward_attr,
+                 max_generation=2,
+                 population_size=10,
+                 keep_top_ratio=0.2,
+                 selection_bound=0.4,
+                 crossover_bound=0.6):
+        """
+        Initialize GeneticSearcher.
+
+        Args:
+            search_space (SearchSpace): The space to search.
+            reward_attr: The attribute name of the reward in the result.
+            max_generation: Max iteration number of genetic search.
+            population_size: Number of trials of the initial generation.
+            keep_top_ratio: Ratio of the top performance population.
+            selection_bound: Threshold for performing selection.
+            crossover_bound: Threshold for performing crossover.
+        """
+        super(GeneticSearch, self).__init__(search_space, reward_attr)
+
+        self._cur_generation = 1
+        self._max_generation = max_generation
+        self._population_size = population_size
+        self._keep_top_ratio = keep_top_ratio
+        self._selection_bound = selection_bound
+        self._crossover_bound = crossover_bound
+
+        self._cur_config_list = []
+        self._cur_encoding_list = []
+        for _ in range(population_size):
+            one_hot = self.search_space.generate_random_one_hot_encoding()
+            self._cur_encoding_list.append(one_hot)
+            self._cur_config_list.append(
+                self.search_space.apply_one_hot_encoding(one_hot))
+
+    def _select(self):
+        population_size = len(self._cur_config_list)
+        logging.info("The %sth generation, population=%s"
+                     % (self._cur_generation, population_size))
+        return self._cur_config_list, self._cur_encoding_list
+
+    def _feedback(self, trials):
+        self._cur_generation += 1
+        if self._cur_generation > self._max_generation:
+            return AutoMLSearcher.TERMINATE
+
+        sorted_trials = sorted(
+            trials, key=lambda t: t.best_result[self.reward_attr], reverse=True)
+        self._cur_encoding_list = self._next_generation(sorted_trials)
+        self._cur_config_list = []
+        for one_hot in self._cur_encoding_list:
+            self._cur_config_list.append(
+                self.search_space.apply_one_hot_encoding(one_hot))
+
+        return AutoMLSearcher.CONTINUE
+
+    def _next_generation(self, sorted_trials):
+        """Generate genes (encodings) for the next generation.
+
```

Comment:
Here, can you provide more details on how the algorithm is implemented?

Reply:
done